File: vera-backend/apps/agent_worker/src/agent_worker/dtmf.py
"""DTMF (keypad tone) sending, isolated from the agent/tool plumbing for testability.

The IVR navigator presses phone-menu digits over the LiveKit SIP call:
`LocalParticipant.publish_dtmf` emits a SIP DTMF message that the livekit-sip service
relays to the PSTN leg. Kept DB/agent-free so it unit-tests against a fake participant.
"""

import asyncio
import logging

from livekit import rtc

logger = logging.getLogger("agent_worker")

# Keypad char -> RFC 4733 DTMF event code: 0-9 -> 0-9, * -> 10, # -> 11, A-D -> 12-15.
_DTMF_CODE: dict[str, int] = {
    **{str(d): d for d in range(10)},
    "*": 10,
    "#": 11,
    "A": 12,
    "B": 13,
    "C": 14,
    "D": 15,
}


class InvalidDtmfError(ValueError):
    """A requested keypad sequence contains an unsupported character."""


class DtmfTransportError(Exception):
    """Publishing a DTMF tone over the live call failed at the SIP/transport layer.

    Wraps the underlying SDK error so callers handle one DTMF-owned type and never
    reach into livekit's internals."""


async def send_dtmf(participant: rtc.LocalParticipant, digits: str, *, gap_s: float = 0.15) -> str:
    """Send each character of `digits` as a SIP DTMF tone, in order, with a short gap.
    Returns the normalized sequence actually sent (stripped/uppercased).

    Validates the whole sequence first and raises `InvalidDtmfError` before sending
    anything, so a bad character never emits a partial sequence. A publish failure on
    the live call is wrapped as `DtmfTransportError`."""
    seq = digits.strip().upper()
    if not seq:
        # An empty sequence emits zero tones; reject it so a caller can't mistake the
        # silent no-op for a successful press.
        raise InvalidDtmfError("empty DTMF sequence")
    bad = sorted({c for c in seq if c not in _DTMF_CODE})
    if bad:
        raise InvalidDtmfError(f"unsupported DTMF characters: {bad}")
    n = len(seq)
    # Count only — never log the raw digit sequence (it can be PHI, e.g. a member ID).
    logger.debug("send_dtmf: emitting %d DTMF tone(s)", n)
    for i, c in enumerate(seq):
        try:
            await participant.publish_dtmf(code=_DTMF_CODE[c], digit=c)
        except Exception as exc:
            raise DtmfTransportError(str(exc)) from exc
        if i < n - 1:
            await asyncio.sleep(gap_s)
    return seq

File: vera-frontend/src/lib/voice-lab/dtmf.ts
// Send DTMF (keypad tones) from the browser into the live LiveKit call. The local
// participant publishes SIP DTMF; livekit-sip relays it to the PSTN leg. Kept free of
// React/livekit-client types so it unit-tests against a fake publisher. Mirrors the
// backend agent_worker/dtmf.py (keep the code map in sync).

/** Keypad char -> RFC 4733 DTMF event code. The Keypad UI offers only 0-9, * and #. */
export const DTMF_CODE: Record<string, number> = {
  "0": 0,
  "1": 1,
  "2": 2,
  "3": 3,
  "4": 4,
  "5": 5,
  "6": 6,
  "7": 7,
  "8": 8,
  "9": 9,
  "*": 10,
  "#": 11,
}

export class InvalidDtmfError extends Error {
  constructor(badChars: string[]) {
    super(`unsupported DTMF characters: ${JSON.stringify(badChars)}`)
    this.name = "InvalidDtmfError"
  }
}

/** The slice of LocalParticipant we need — a LocalParticipant satisfies this. */
export interface DtmfPublisher {
  publishDtmf(code: number, digit: string): Promise<void>
}

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))

/**
 * Send each character of `digits` as a SIP DTMF tone, in order, with a short gap.
 * Validates the whole sequence first and throws `InvalidDtmfError` before sending
 * anything, so a bad character never emits a partial sequence.
 */
export async function sendDtmf(
  publisher: DtmfPublisher,
  digits: string,
  { gapMs = 150 }: { gapMs?: number } = {},
): Promise<void> {
  const seq = digits.trim()
  const bad = [...new Set([...seq].filter((c) => !(c in DTMF_CODE)))]
  if (bad.length) throw new InvalidDtmfError(bad)
  for (let i = 0; i < seq.length; i++) {
    const digit = seq[i]
    const code = DTMF_CODE[digit]
    console.debug("[DTMF] publish", { digit, code })
    await publisher.publishDtmf(code, digit)
    if (i < seq.length - 1) await sleep(gapMs)
  }
  console.info("[DTMF] sent %d tone(s): %s", seq.length, seq)
}
